src/main/java/dev/jbang/cli/Build.java

```diff
@@ -17,7 +17,9 @@ public Integer doCall() throws IOException {
 		RunContext ctx = getRunContext();
 		Code code = ctx.forResource(scriptMixin.scriptOrFile);
 
-		buildIfNeeded(code, ctx);
+		if (code.needsBuild(ctx)) {
```

Comment:
why not have this in the method to avoid the 2-3x repetition?

Reply:
It's basically because I'm not sure we want this to be too "black box". You would have no proper way to detect if a build is possibly going to be done or not. But it could definitely be done.

Reply:
Ah yes, now that I try to change it I remember why it's like this :-)

A `Builder` takes a `SourceSet` and turns it into a `Jar`, which is executable. But in case of JShell scripts it's actually the `SourceSet` itself which can be executed directly so no build is needed.

So right now we can do:

```java
Code code = ...
if (code.needsBuild()) {
   // It's java, kotlin or groovy code here
   code = code.builder().build();
   // Now it's a Jar
} else {
   // It's Jshell code here
}
// execute "code", whatever type it is
```

So of course we can change `Builder.build()` to return `Code` and just make it a no-op to try to build a `SourceSet` containing JShell code ands simply return the `SourceSet`.

But it feels a bit weird to do that. We'd be making the API slightly worse just to save a couple of lines of code. 
I had it like this at some point but decided against it and rewrote it to what it is now. But we can always change our minds again :-)

Reply:
Ok, the changes aren't too big so I added an extra commit with these changes. Let's see if it's nicer or not.

Reply:
Well actually to get things to work correctly I did have to make a bunch more changes. But I'm actually pretty happy with it. I also added new feature where you can force a source to be of any type, not just Jshell. For that purpose I added a `--source-type` option and deprecated `--jsh` (adding a bunch of `--kotlin`, `--groovy` options was out of the question of course).